### ISSUE TITLE
Thinking about containerization

### ISSUE THREAD

**Issue by vsoch** [prefix-dev/pixi#634]
### Problem description

There might be a better issue (or discussion) to post this, but I'm thinking through a few recent cases of building containers with pixi and want to share the features that I would like.

- **A container base that boots up and I have pixi ready to go**. Yes, it's trivial to curl to install, but in practice what I'm doing is curling, then creating the environment, and then all my subsequent commands need to source the environment first (and when I shell in or do an entrypoint I have to do the same). I ideally would like a setup where I have pixi, and I can either build or shell and just get to work (and not need to source the bash profile, find the environment, etc.)
- **A building workflow that encourages me to save the pixi.lock**: For most of my environments, I have the pixi.toml, add it to the container, and build. Most of the time this works fine to build and build again. Of course I was burned when my previous pixi.toml would not reproduce. I pulled an old container, grabbed the lock file, and was able to fix it. But really a better approach would be to get that file on the host. This one depends on what you all are thinking for pixi with build - if you are going to do a wrapper sort of deal (like spack containerize but actually running a build command) you could copy the lock file (if it does not exist). If you just dump out the Dockerfile, you can have a suggested command at the end to copy it to the machine post build (and this is probably simplest and a good first shot).

Those are the two things on my mind atm - back to building! :package:

**Comment by ruben-arts** [comment]
He @vsoch 

- Building a base container: fair point we should build one for the users!
- Encourages me to save the `pixi.lock`: You can do it!! ( I'm not sure what you question is here :wink: )
- Environment variable for root:  You can do `pixi add --manifest-path ~/vsoch/magical-pkg/pixi.toml` but that seems to be even more work then the `cd'`ing. We currently don't want to support "global" environments as we think its a better experience to have environments linked to an actual path. Maybe this will change in the future but not in the near one.

**Comment by vsoch** [comment]
Hi @ruben-arts !

> Encourages me to save the pixi.lock: You can do it!! ( I'm not sure what you question is here 😉 )

Most will start with a pixi.toml on the host, and the pixi.lock is generated in the container. They won't think to save it (copying it out of the container post build) and this is especially true for automated builds.

> Environment variable for root: You can do pixi add --manifest-path ~/vsoch/magical-pkg/pixi.toml but that seems to be even more work then the cd'ing. We currently don't want to support "global" environments as we think its a better experience to have environments linked to an actual path. Maybe this will change in the future but not in the near one.

Yes, that's what I sometimes do, but it's annoying to type all that. Perhaps an intermediate suggestion is to require an environment, but allow setting `PIXI_MANIFEST_PATH` so that when I run `pixi <command>` if there is no pixi.toml in the present working directory, it looks for that environment variable. That way I'd still be exec'ing commands to the environment but I could do so from anywhere in the container. imho that is a much better development experience.

Ping @wolfv for his thoughts!

**Comment by ruben-arts** [comment]
I'm not sure what we can do about the lockfile staying in the docker image. This is more a workflow then a pixi issue right? 

I normally run the solve locally and then when i'm content with the project I will move the repo into an image and build it into a docker image. But I guess your use-case is that your environment can not be build locally, right?

**Comment by vsoch** [comment]
The issue is that you need the lock file in the outside of the container to copy to it to ensure the same thing builds.

**Comment by wolfv** [comment]
Yeah I think we all agree that we'd love to have a lean and mean pixi base image that contains only pixi and some slim base (like `micromamba-docker`).

We could have the pixi environment be default mounted to something like `/pixi` and the entry point would be `pixi shell --manifest /pixi/` or something like that. 

We would want to copy both the `pixi.toml` and the `pixi.lock` file into that folder.

With micromamba, it's maybe a bit easier because usually the user only needs a single file to "recreate" the env (that is the `environment.yml` file. With pixi we'd probably advise the users to take both, the `.toml` and `.lock` file. 

We could also create some template with pixi to create a boilerplate `Dockerfile`.

Anyways, just a bit of a braindump from my side :)

**Comment by brad-jones** [comment]
This could end up being the base for https://github.com/brad-jones/pixi-devcontainer

**Comment by vsoch** [comment]
Absolutely! And I'll try them out the next time around. Here are the images in GitHub packages: https://github.com/prefix-dev/pixi-docker/pkgs/container/pixi.

Thanks @ruben-arts !

**Comment by pavelzw** [comment]
Unfortunately the docs are not quite there yet for docker. But here is an example of how to use it: https://github.com/prefix-dev/pixi/tree/main/examples/docker
